extensions/amp-story-shopping/0.1/test/test-amp-story-shopping-config.js

```diff
@@ -131,5 +154,102 @@ describes.realWin(
           .to.have.been.calledOnce;
       });
     });
+
+    it('test validate required fields for config', async () => {
+      const invalidConfig = {
+        'items': [{}],
+      };
+
+      await createAmpStoryShoppingConfig(null, invalidConfig);
+
+      for (const [key, value] of Object.entries(PRODUCT_VALIDATION_CONFIG)) {
+        if (value.includes(validateRequired)) {
+          expect(() => {
+            validateRequired(key, invalidConfig['items'][0][key]);
+          }).to.throw(`Field ${key} is required.`);
+        }
+      }
+    });
+
+    it('test config is a string', async () => {
+      const invalidConfig = {
+        'items': [
+          {
+            'productId': 'city-pop',
+            'productTitle': 123,
+            /* Not a string, will fail string type test */
+            'productPrice': 19,
+            'productPriceCurrency': 'JPY',
+            'productImages': [
+              '/examples/visual-tests/amp-story/img/shopping/nest-mini-icon.png',
+              '/examples/visual-tests/amp-story/img/shopping/nest-mini-icon.png',
+            ],
+          },
+        ],
+      };
+
+      await createAmpStoryShoppingConfig(null, invalidConfig);
+      expect(() => {
+        validateString(
+          'productTitle',
+          invalidConfig['items'][0]['productTitle']
+        );
+      }).to.throw(
+        `productTitle ${invalidConfig['items'][0]['productTitle']} is not a string.`
+      );
+    });
```

Comment:
Do these tests work? Anything before the `expect` line 192 seems to be unused. Why do you need to call the validation function manually? It's supposed to be running as part of the `createAmpStoryShoppingConfig` already.

Reply:
Yes, the tests work. The validation functions are not part of the component. We discussed in the office hour that we might want to take an alternative approach as it was difficult to spy/stub given that we have these functions outside of the component. Also, not sure what you mean by unused, I am using invalidConfig['items'][0]['productTitle'] as the second parameter in the validateString function.

Reply:
The goal of a unit test is to make sure everything works. The best way to accomplish that is to replicate the exact code path users will run into, by calling the `entry door` to your component, and verifying that what was supposed to happen happened. In that case it seems to be `createAmpStoryShoppingConfig`.

If you remove the line that runs all your code `await createAmpStoryShoppingConfig(null, invalidConfig);`, your test still passes, so what does it test? Only the `validateString` helper, but it doesn't test your component at all.

You need to let the code that validates the config run, and then verify that it did what it was supposed to do. Does it call a method? Does it output some logs? Does it set a class? This is what you need to verify, but if you call the validation function yourself while your code is supposed to do it, your test becomes different from what users will run in their browser, and is therefore of very little use.

Reply:
Ok, I see what you mean. I'll try to stub the calls to the validation functions before `createAmpStoryShoppingConfig`, to make sure that the correct validation functions get called and the same code path that the user takes is covered with the tests. I think it is beneficial to keep the helper functions' tests as they also need to work properly.

Reply:
I added in some spies on the helper functions with the error config param items. Calling `createAmpStoryShoppingConfig` will trigger the complete codepath that the user will go through using the public API via `getShoppingConfig`. These tests should be good to go now. Let me know if you need anything else.